### PR TITLE
Id cleanup

### PR DESCRIPTION
**Pull request by jeromekelleher**
This is a fairly widespread refactoring of the datamodel code. Changes:

1. All of the datamodel objects actually inherit from datamodel.DatamodelObject now. Each one has a CompoundId and a localId. 
2. Refactored the CompoundId concept a bit to fit into this.

The aim of this is to get a proper, linked up model in memory of our data tree. This should make issues like #535 simple to fix (should be fixed already, in fact; it would be better to verify this and add some test cases in a follow up PR though).

There's still some tests to fix, but I ran out of time. Hopefully I'll finish this PR later.

Comments/feedback?

### CODE REVIEW

**Comment by jeromekelleher** [comment]
This is ready for review. I haven't squashed the commits yet as I can't decide where the best place for the CompoundId classes is. It seems like they belong in the specific modules (which is where they are now), but his creates a circular import from datasets->reads/variants->datasets. I've worked around this issue fairly crudely and had to remove the snakefood test for circular imports. It works fine, but it's a bit ugly. The other approach is to put all the CompoundId classes into datamodel, which avoids this problem but seems confusing. 

Anyway, comments appreciated.

**Comment by dcolligan** [comment]
Circular imports should be avoided unless there is a really, really good reason and this is not one of them.  We should prefer the solution that avoids circular imports.

**Comment by dcolligan** [comment]
Okay, in general I like this refactor, although it does tie us strongly into hierarchical IDs.  My previous classes were a bit more flexible in that regard.  I suppose this is more appropriate for the data model we have now, however.

**Comment by jeromekelleher** [comment]
Excellent, glad to hear it. I'm happy to change the circular import stuff. My original approach was to have all of the CompoundId classes in the `datamodel` module, which avoided the circular imports. See https://github.com/jeromekelleher/server/commit/ee0eb42d55b6733c86211554de9d8a9572f6ceba to get an idea of the changes. Are you happy reverting to something like this, or is there a better option?

File: ga4gh/datamodel/datasets.py [hunk 1]
```diff
@@ -97,30 +94,32 @@ class SimulatedDataset(AbstractDataset):
...
+            self, localId, randomSeed=1, numCalls=1,
+            variantDensity=1, numVariantSets=1, numAlignments=1):
+        super(SimulatedDataset, self).__init__(localId)
         self._randomSeed = randomSeed
         self._randomGenerator = random.Random()
         self._randomGenerator.seed(self._randomSeed)
 
+        # We do the imports here because we have circular dependencies
+        # otherwise.
+        import ga4gh.datamodel.variants as variants
```
Comment (dcolligan) [inline ga4gh/datamodel/datasets.py]: ew ew ew ew ew ew!  Let's move all the CompoundId subclasses to datamodel/__init__.py

File: ga4gh/datamodel/datasets.py [hunk 2]
```diff
@@ -131,34 +130,33 @@ class FileSystemDataset(AbstractDataset):
...
-        super(FileSystemDataset, self).__init__()
-        self._id = os.path.basename(os.path.normpath(datasetDir))
-        self._datasetDir = datasetDir
+    def __init__(self, dataDir):
+        localId = os.path.basename(os.path.normpath(dataDir))
+        super(FileSystemDataset, self).__init__(localId)
+
+        # We do the imports here because we have circular dependencies
+        # otherwise.
+        import ga4gh.datamodel.variants as variants
```
Comment (dcolligan) [inline ga4gh/datamodel/datasets.py]: same here

File: ga4gh/datamodel/reads.py [hunk 3]
```diff
@@ -11,37 +11,34 @@
...
 
-import ga4gh.protocol as protocol
 import ga4gh.datamodel as datamodel
+import ga4gh.datamodel.datasets as datasets
 import ga4gh.exceptions as exceptions
+import ga4gh.protocol as protocol
 
 
-class CompoundReadGroupSetId(datamodel.CompoundId):
+class ReadGroupSetCompoundId(datasets.DatasetCompoundId):
```
Comment (dcolligan) [inline ga4gh/datamodel/reads.py]: name change to have the 'compound' part after the class type is probably a good idea

File: ga4gh/datamodel/reads.py [hunk 4]
```diff
@@ -11,37 +11,34 @@
...
     """
     The compound id for a read group set
     """
-    fields = ['datasetId', 'rgsId']
+    fields = datasets.DatasetCompoundId.fields + ['readGroupSet']
```
Comment (dcolligan) [inline ga4gh/datamodel/reads.py]: composing the fields from the parent class' field attr is a good idea

File: tests/unit/test_imports.py [hunk 5]
```diff
@@ -35,10 +35,6 @@ def setUpClass(cls):
         snakefoodScanner = SnakefoodScanner()
         cls.graph = snakefoodScanner.scan()
 
-    def testNoCycles(self):
```
Comment (dcolligan) [inline tests/unit/test_imports.py]: reenable this

**Comment by dcolligan** [comment]
yeah, put CompoundId classes in the datamodel module

**Comment by jeromekelleher** [comment]
@dcolligan --- ickiness removed. Hopefully ready to land?

**Comment by dcolligan** [comment]
+1

**Pull request merged**